### ISSUE TITLE
The ReadStringAsCharset/WriteStringAsCharset methods work incorrectly and break content

### ISSUE THREAD

**Issue by evgeny-k** [IndySockets/Indy#387]
Reproduced with Indy shipped with `Delphi 10.4.2`, `Delphi 11`. 
other versions of Delphi aren't tested.

steps:
* Launch
* see error:
* `Project Project5.exe raised exception class Exception with message 'Size is differ. expect: 49 actual: 65'.`

simple testcase:

```
program Project5;

{$APPTYPE CONSOLE}

{$R *.res}

uses
  System.SysUtils, System.Classes,
  idGlobalProtocols;

const
  c_str: array [0..48] of Byte =
       ($52, $4F, $31, $30, $37, $00, $01, $00, $00, $00,
        $00, $00, $39, $C9, $59, $5B, $B1, $96, $68, $42,
        $85, $C6, $94, $68, $F8, $8F, $F8, $B8, $0C, $00,
        $00, $00, $45, $52, $4F, $45, $78, $63, $65, $70,
        $74, $69, $6F, $6E, $01, $00, $00, $00, $3F);

var
  mem, mem1: TMemoryStream;
  str: String;
begin
  mem := TMemoryStream.Create;
  mem.Write(c_str, Length(c_str));
  mem1 := TMemoryStream.Create;
  mem.Position := 0;
  try
    try
      writeln('original stream size = '+mem.Size.ToString);
      str := ReadStringAsCharset(mem, 'utf-8');
      writeln('size after reading = '+(Length(str)).ToString);
      WriteStringAsCharset(mem1,str,'utf-8');
      writeln('stream size after writing = '+mem1.Size.ToString);
      if mem1.Size <> Mem.Size then
        raise Exception.Create('Size is differ. expect: '+mem.Size.ToString +' actual: '+mem1.Size.ToString);
    finally
      mem.Free;
      mem1.Free;
    end;
  except
    on E: Exception do
      Writeln(E.ClassName, ': ', E.Message);
  end;
end.


```

**Comment by rlebeau** [comment]
The test data you are using is not fully textual, it contains non-text binary bytes in it, too.  But, while that in of itself will not break UTF-8 processing, more importantly is the fact that your test data is simply not valid UTF-8 to begin with, there are many errors in it, specifically:

- `$C9` starts a 2-byte sequence, but is followed by `$59` which is not a valid continuation byte.
- `$B1`, `$96`,  `$85`, `$8F`, `$B8`, these are all rogue continuation bytes that are not preceded by any lead bytes.
- `$F8` is an illegal byte in UTF-8.

So, I would expect `ReadStringAsCharSet()` to return a badly decoded `string` (since it does not raise an exception on failure), and then because of that `WriteStringAsCharSet()` will create a new set of UTF-8 bytes that are different than you are expecting.

Though, I would not expect the final result to be _larger_ than the original input, since decoding bad data _should_ be producing fewer `string` characters (specifically, bad byte sequences should be decoded to character `'?'`), and thus encode to fewer bytes (namely, `$3F`).

**Comment by evgeny-k** [comment]
ReadStringAsCharSet reads binary data as `'RO107'#0#1#0#0#0#0#0'Ę��/�M�n�~j�'#$B'7'#$C#0#0#0'EROException'#1#0#0#0'?'`.  
this part (`'Ę��/�M�n�~j�'#$B'7'`) is Guid (16 bytes) so it can be new for each response.

My purpose is restore original stream and get this binary data. However I can't get raw stream because you have destroyed it in `CheckException` method, so I'm trying to get original stream via encoding it with `WriteStringAsCharset` ...
```
function TIdHTTPProtocol.ProcessResponse(AIgnoreReplies: array of Int16): TIdHTTPWhatsNext;
...
  procedure CheckException;
...
        except
          on E: Exception do begin
            FHTTP.Disconnect;
            if not (E is EIdConnClosedGracefully) then begin
              raise;
            end;
          end;
        end;
        if LRaiseException then begin
          LTempStream.Position := 0;
          raise EIdHTTPProtocolException.CreateError(LResponseCode, Response.ResponseText,
            ReadStringAsCharset(LTempStream, Response.CharSet));// <<< Response.CharSet = 'utf-8'
...
    finally
      if LRaiseException then begin
        LTempStream.Free; //<<<
      end;
    end;
  end;

```

**Comment by evgeny-k** [comment]
original issue was solved in #388

**Comment by rlebeau** [comment]
> ReadStringAsCharSet reads binary data as `'RO107'#0#1#0#0#0#0#0'Ę��/�M�n�~j�'#$B'7'#$C#0#0#0'EROException'#1#0#0#0'?'`. this part (`'Ę��/�M�n�~j�'#$B'7'`) is Guid (16 bytes) so it can be new for each response.

That is not textual data, and you certainly cannot have a binary GUID in a Unicode string. So this data should not be decoded as UTF-8 to begin with.  That is not a fault in `ReadStringAsCharSet()`.

> My purpose is restore original stream and get this binary data.

That is not going to happen in this case, since the resulting `string` is corrupted, so you can't reverse it back into the original bytes.  The best solution would be to handle the original bytes as-is and not decode them as UTF-8 at all.

> However I can't get raw stream because you have destroyed it in `CheckException` method

You are referring to code in `TIdHTTP` that processes an HTTP response and then raises `EIdHTTPProtocolException` on an HTTP error, where the exception's `ErrorMessage` property is the content of the response in `string` format.  That property is not intended to handle binary data, as it is not typical for HTTP errors to carry binary data.

> so I'm trying to get original stream via encoding it with `WriteStringAsCharset` ...

You are going about this the wrong way.  To get the original bytes of an HTTP error, don't use `EIdHTTPProtocolException` at all.  You can instead enable the `hoNoProtocolErrorException` and `hoWantProtocolErrorContent` flags in the `TIdHTTP.HTTPOptions` property, and then use you own `TStream` to receive the raw bytes from `TIdHTTP`, eg:

```
IdHTTP1.HTTPOptions := IdHTTP.HTTPOptions + [hoNoProtocolErrorException, hoWantProtocolErrorContent];
IdHTTP1.Get(..., MyTStream);
// use MyTStream as needed...
```
